src/base/locked_queue.h

```diff
@@ -56,41 +63,130 @@ class LockedQueue {
   }
 
   // Returns disengaged object only when this queue is closed and empty.
-  Optional Pop() THREAD_SAFE {
+  Optional Pop(const ui32 shard = 0u) THREAD_SAFE {
+    return Pop(nullptr, shard);
+  }
+
+  Optional Pop(Atomic<ui64>* external_counter,
+               const ui32 shard = 0u) THREAD_SAFE {
     UniqueLock lock(pop_mutex_);
     pop_condition_.wait(lock, [this] { return closed_ || !queue_.empty(); });
     if (closed_ && queue_.empty()) {
       return Optional();
     }
-    Optional&& obj = std::move(queue_.front());
-    queue_.pop();
+
+    auto task = index_.Get(shard);
+
+    Optional&& obj = std::move(*task);
+    queue_.erase(task);
     --size_;
+    if (external_counter) {
+      ++(*external_counter);
+    }
     return std::move(obj);
   }
 
-  Optional Pop(Atomic<ui64>& external_counter) THREAD_SAFE {
+  void UpdateIndex(const ui32 shards_number) THREAD_SAFE {
+    CHECK(shards_number > 0u);
     UniqueLock lock(pop_mutex_);
-    pop_condition_.wait(lock, [this] { return closed_ || !queue_.empty(); });
-    if (closed_ && queue_.empty()) {
-      return Optional();
-    }
-    Optional&& obj = std::move(queue_.front());
-    queue_.pop();
-    --size_;
-    ++external_counter;
-    return std::move(obj);
+    index_.Update(shards_number);
   }
 
  private:
   friend class QueueAggregator<T>;
 
+  class Index {
+   public:
+    using TaskInfo = Tuple<typename Queue::iterator, size_t>;
+    using ShardsList = List<TaskInfo>;  // List of tasks from same shard.
+    enum InfoIndex {
+      ITERATOR = 0u,
+      HASH = 1u,
+    };
+    enum ReverseInfoIndex {
+      RANGE = 0u,
+      REVERSE_ITERATOR = 1u,
+    };
+
+    Index(LockedQueue<T>* const queue) : queue_(queue) {
+      CHECK(queue_) << "Queue must be specified";
+    }
+
+    ~Index() {
+      // Make sure we don't leak tasks.
+      DCHECK(reverse_index_.size() == Size());
+    }
+
+    size_t Size() const THREAD_UNSAFE {
+      size_t index_size = 0u;
+      for (const ShardsList& shards_list : index_) {
+        index_size += shards_list.size();
+      }
+      return index_size;
+    }
+    void Put(typename Queue::iterator item, const ui64 hash) {
```

Comment:
Actually, no. Here and in other places we hash is a hash value computed out of sources. Base on hash value and number of shards a shard could be computed. But anyway we need 2 distinct entities - shards and hashes.